ostd/src/cpu/set.rs

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module contains the implementation of the CPU set and atomic CPU set.
+
+use core::sync::atomic::{AtomicU64, Ordering};
+
+use smallvec::SmallVec;
+use static_assertions::const_assert_eq;
+
+use super::{num_cpus, CpuId};
+
+/// A subset of all CPUs in the system.
+#[derive(Clone, Debug, Default)]
+pub struct CpuSet {
+    // A bitset representing the CPUs in the system.
+    bits: SmallVec<[InnerPart; NR_PARTS_NO_ALLOC]>,
```

Comment:
If I understand, the use of `SmallVec` is for performance, right? I noticed it involves some unsafe things. I suggest we may use it when we observe a performance bottleneck here. But please let me know if it's not the case. Same for `AtomicCpuSet`.

Reply:
Yes it is for performance. Also `CpuSet` is used on remote function calls so it is better to be fast. We have many dependencies that includes "unsafe" usage but I think it's quite a question if all the dependencies that uses unsafe code for implementation can't be trusted. Had it lead to serious vulnerability incidents in the past? It looks like that it can be trusted.

Reply:
>  I think it's quite a question if all the dependencies that uses unsafe code for implementation can't be trusted.

I think, which you may disagree, logically, any unsafe code not maintained by us should be seen as vulnerable and we should be careful. But I'm not against the use of this kind of dependencies if we have to.

> Had it lead to serious vulnerability incidents in the past? It looks like that it can be trusted.

If you mean the `smallvec` crate, the answer may be "no" except for two known CVEs. And it's kind of widly used in Rust community. So I'm OK with the usage if you insist since you know far more than me about `CpuSet`'s use cases.

Reply:
We are already using `smallvec` for `XArray`. In the future we may implement dynamically sized `CpuSet` without heap allocation to replace it since the `CpuSet` can always have a fixed size after booting. I think it is a matter of expediency to use `smallvec` for `CpuSet` for now.